### ISSUE TITLE
Type error when calling methods of protocol superclass

### ISSUE THREAD

**Issue by medihack** [microsoft/pyright#7160]
**Describe the bug**

In previous versions of pyright the below code didn't have any type errors. 

```python
class Base(Protocol):
    def get_data(self):
        ...

class MyMixin:
    def get(self: Base):
        self.get_data()
        super().get_data()
```

But at least since v1.1.348 (and also the current v1.1.348) the last line (`super().get_data()`) leads to the error (as mentioned in the [discussions](https://github.com/microsoft/pyright/discussions/6343#discussioncomment-8287672)):

```
error: Method "get_data" cannot be called because it is abstract and unimplemented (reportAbstractUsage)
```

**VS Code extension or command-line**

Currently running Pyright v1.1.349 from the command line.

**Comment by erictraut** [comment]
Thanks for the bug report. When you initially mentioned this in [the discussion](https://github.com/microsoft/pyright/discussions/6343#discussioncomment-8247351), I assumed it was related to [this issue](https://github.com/microsoft/pyright/issues/7105), but that was an incorrect assumption.

This issue is related to a [different change](https://github.com/microsoft/pyright/issues/6880) that was required for conformance with the typing spec.

Your use case is a little different than the case covered by the typing spec because you've explicitly declared the type of the `self` parameter to be a protocol. (Technically, the type of `self` should be `Base & Mixin` where `&` represents an intersection, but intersections are not yet supported by the type system.)

I'll need to think more about how to implement this check in a way that is both conformant with the typing standard and enables your use case.

**Comment by medihack** [comment]
Getting type hints correctly for Mixins in Django is a real pain (and this is not just my use case). And yes, I came to the same conclusion that intersections [are missing in the Python type system](https://github.com/python/typing/issues/213). Especially when the Mixin itself has itself attributes, here is a real-world example:

```python
class ViewProtocol(Protocol):
    request: HttpRequest

    def dispatch(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
        ...

    def get(self, request: HttpRequest, *args: Any, **kwargs: Any) -> HttpResponse:
        ...

    def get_context_data(self, **kwargs: Any) -> dict[str, Any]:
        ...


class LockedMixinProtocol(ViewProtocol, Protocol):
    settings_model: type[AppSettings]
    section_name: str


class LockedMixin:
    def dispatch(
        self: LockedMixinProtocol, request: HttpRequest, *args: Any, **kwargs: Any
    ) -> HttpResponse:
        settings = self.settings_model.get()

        if settings.locked and not (is_logged_in_user(request.user) and request.user.is_superuser):
            if request.method != "GET":
                raise SuspiciousOperation()

            return TemplateView.as_view(
                template_name="core/section_locked.html",
                extra_context={"section_name": self.section_name},
            )(request)

        return super().dispatch(request, *args, **kwargs)
```

I hope there will be a solution (I really hate those `# type: ignore` ;-)). It is also the way [mypy recommends it](https://mypy.readthedocs.io/en/stable/more_types.html#mixin-classes), but I have not tried it there as I completely switched over to pyright.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.350, which I just published.
